packages/dart_frog/lib/src/provider.dart

```diff
@@ -1,10 +1,90 @@
+import 'dart:async';
+
 import 'package:dart_frog/dart_frog.dart';
 
-/// Provide a value to the current handler by calling [create] lazily.
+final _cache = <Type, Object>{};
+
+/// Clears the in-memory provider cache.
+void clearProviderCache() => _cache.clear();
+
+/// Provide an object to the current handler by calling [create].
+///
+/// If [lazy] is `true`, [create] will be called only when performing
+/// a lookup via `context.read`.
+/// If [lazy] is `false`, [create] will be called immediately when
+/// an incoming request is received regardless of if a lookup is performed.
+///
+/// By default all providers are [lazy], meaning [create] is called on demand.
+///
+/// {@template provider_cache}
+/// If [cache] is `true`, [create] will only be called once and the result
+/// will be cached for the lifetime of the application.
+/// If [cache] is `false`, [create] will be called multiple times and
+/// the provided object will be recomputed.
+///
+/// By default providers cache, meaning the provided object won't be recomputed.
+/// {@endtemplate}
+///
+/// ```dart
+/// Handler middleware(Handler handler) {
+///  return handler.use(provider<String>((_) => 'Hello World!'));
+/// }
+/// ```
 Middleware provider<T extends Object>(
-  T Function(RequestContext context) create,
-) {
+  T Function(RequestContext context) create, {
+  bool cache = true,
+  bool lazy = true,
+}) {
+  T _create(RequestContext context) {
+    return cache ? _memo<T>(() => create(context)) : create(context);
+  }
+
   return (handler) {
-    return (context) => handler(context.provide(() => create(context)));
+    return (context) {
+      if (lazy) return handler(context.provide(() => _create(context)));
+      final value = _create(context);
+      return handler(context.provide(() => value));
+    };
   };
 }
+
+/// Provide a `Future<T>` to the current handler by calling [create].
+/// [create] will be called when an incoming request is received.
+///
+/// {@macro provider_cache}
+/// ```dart
+/// Handler middleware(Handler handler) {
+///  return handler.use(
+///    futureProvider<String>((_) async => 'Hello World!'),
+///  );
+/// }
+/// ```
+Middleware futureProvider<T extends Object>(
+  Future<T> Function(RequestContext context) create, {
+  bool cache = true,
+}) {
+  Future<T> _create(RequestContext context) async {
+    return cache ? _asyncMemo<T>(() => create(context)) : create(context);
+  }
+
+  return (handler) {
+    return (context) async {
+      final value = await _create(context);
```

Comment:
If I understand correctly, does this means the objects will be created before the hander is called? If so, this doesn't really work for my particular use case since we have a lot of routes that don't need certain expensive dependencies and I'd rather not build them if we can avoid it (although the caching will help a lot with that).

I suppose we could just rework our routes so that the middleware is only provided when needed, but then we'd have to design our route structure based on functionality, not what makes logical sense (following REST for example).

If there was an optional parameter that was like `bool Function(RequestContext context) createWhen` so I could tell the provider "_only create this expensive dependency if a certain header is set_" , I think that would work well too. It may be a bit convoluted for beginners, but it'd be highly optional 🤷

Reply:
> If I understand correctly, does this means the objects will be created before the hander is called?

In that case you could use a regular, lazy, cached provider and provide/read the `Future`. That would result in the value being computed only one time and only when it is read:

```dart
Handler middleware(Handler handler) {
  return handler.use(provider<Future<AsyncValue>>((_) => _createAsyncValue());
}
```

Then you can access it via:

```dart
Future<Response> onRequest(RequestContext context) async {
  final asyncValue = await context.read<Future<AsyncValue>>();
} 
```

Hope that helps!

Reply:
Would that cache the actual value or the `Future` of that value (since in that case it'd be using `_memo` and not `_asyncMemo`)? If it is caching the `Future`, would it recompute each time it's accessed?

Reply:
It should cache the `Future` so the computation would only done once.